### PR TITLE
[stm] implementation of audiopwmio

### PR DESCRIPTION
**Pull request by tyomitch**
Based on nrf PWMAudioOut by @jepler and stm PulseOut by @hierophect

Tested on a Meowbit

### CODE REVIEW

File: ports/stm/common-hal/microcontroller/Pin.c [hunk 1]
```diff
@@ -194,3 +194,12 @@ void common_hal_mcu_pin_claim(const mcu_pin_obj_t* pin) {
 void common_hal_mcu_pin_reset_number(uint8_t pin_no) {
     reset_pin_number(pin_no / 16, pin_no % 16);
 }
+
+void set_drive_mode(const mcu_pin_obj_t *pin, uint32_t mode) {
```
Comment (hierophect) [inline ports/stm/common-hal/microcontroller/Pin.c:198]: This should be a static function within PWMAudioOut.c, it doesn't need to be in Pin.c. The function in DigitalIO is very old and shouldn't be adjusting the pull and speed as it does, so it needs to be tweaked anyway.

File: ports/stm/common-hal/digitalio/DigitalInOut.c [hunk 2]
```diff
@@ -117,13 +117,8 @@ bool common_hal_digitalio_digitalinout_get_value(
 digitalinout_result_t common_hal_digitalio_digitalinout_set_drive_mode(
         digitalio_digitalinout_obj_t *self,
         digitalio_drive_mode_t drive_mode) {
-    GPIO_InitTypeDef GPIO_InitStruct = {0};
-    GPIO_InitStruct.Pin = pin_mask(self->pin->number);
-    GPIO_InitStruct.Mode = (drive_mode == DRIVE_MODE_OPEN_DRAIN ?
+    set_drive_mode(self->pin, drive_mode == DRIVE_MODE_OPEN_DRAIN ?
```
Comment (hierophect) [inline ports/stm/common-hal/digitalio/DigitalInOut.c:120]: Please revert this as per the previous comment. This module needs some reworks, and Pin.c isn't a good place to abstract this kind of thing in the meantime.

File: ports/stm/common-hal/audiopwmio/PWMAudioOut.c [hunk 3]
```diff
@@ -0,0 +1,350 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2021 Artyom Skrobov for Adafruit Industries
```
Comment (hierophect) [inline ports/stm/common-hal/audiopwmio/PWMAudioOut.c:6]: We use "for Adafruit Industries" to designate work that was sponsored by Adafruit, so you don't have to include it in your byline.

**Review by hierophect — CHANGES_REQUESTED** [review]
Thanks for this PR! I'm unable to get this working on my Meowbit with the WAV sample code in the [Circuitpython documentation](https://circuitpython.readthedocs.io/en/latest/shared-bindings/audiopwmio/index.html). The speaker makes a little strangled sound, and then the program hangs irrecoverably, requiring a manual erase of the entire flash. If you have a test sketch of your own, I'm happy to try that.

The style and timer allocation stuff looks generally ok, I can't point out where the hang might be at first glance. There are also some minor things to fix unrelated to performance.

**Comment by tyomitch** [comment]
@hierophect thank you for the review!

> I'm unable to get this working on my Meowbit with the WAV sample code in the [Circuitpython documentation](https://circuitpython.readthedocs.io/en/latest/shared-bindings/audiopwmio/index.html). The speaker makes a little strangled sound, and then the program hangs irrecoverably, requiring a manual erase of the entire flash. If you have a test sketch of your own, I'm happy to try that.

My guess is that it's #4397 :-)
You should be able to get `WaveFile`s playing if you prepend the call to `play()` with a small `time.sleep()`, so that the screen is no longer updating at the time `PWMAudioOut` tries to access the flash.
The PR replaces the irrecoverable hang with a recoverable failure, so it would simplify testing, but it still wouldn't allow accessing flash in parallel with updating the screen.

`RawSample`s should be playing OK regardless of the SPI locking issue.

**Comment by hierophect** [comment]
@tyomitch ah, thanks for the full context, that probably is it. I'll make a note of my problem for Scott in the issue.

**Comment by hierophect** [comment]
@tyomitch thanks for the updates. Just a gentle note that here on the Circuitpython github, our general policy is to update PRs without force-pushing, so we can see what you've edited and the comments are not lost. It makes it easier to tell when re-testing with hardware is necessary. Is this ready for another review?

**Comment by tyomitch** [comment]
> Is this ready for another review?

Yes please :)

I will try to avoid force-pushing PRs, but for the record, overwritten commits are not lost, and you can see the changes since the first review at https://github.com/adafruit/circuitpython/compare/fc9b42a..b40d072 -- inevitably very noisy because of an intervening global reformatting (a52eb88)

**Review by hierophect — APPROVED** [review]
Tested on Meowbit and Feather, looks good to me!

**Pull request merged**